components/fxa-client/src/internal/close_tabs.rs

```diff
@@ -2,36 +2,91 @@
         decrypt_command, encrypt_command, IncomingDeviceCommand, PrivateCommandKeys,
     },
+    device::COMMAND_MAX_PAYLOAD_SIZE,
     http_client::GetDeviceResponse,
     scopes, telemetry, FirefoxAccount,
 };
 use crate::{Error, Result};
 
 impl FirefoxAccount {
-    pub fn close_tabs<T: AsRef<str>>(&mut self, target_device_id: &str, urls: &[T]) -> Result<()> {
+    pub fn close_tabs<T>(&mut self, target_device_id: &str, urls: Vec<T>) -> Result<()>
+    where
+        T: Into<String>,
+    {
         let devices = self.get_devices(false)?;
         let target = devices
             .iter()
             .find(|d| d.id == target_device_id)
             .ok_or_else(|| Error::UnknownTargetDevice(target_device_id.to_owned()))?;
-        let (payload, sent_telemetry) =
-            CloseTabsPayload::with_urls(urls.iter().map(|url| url.as_ref().to_owned()).collect());
-        let oldsync_key = self.get_scoped_key(scopes::OLD_SYNC)?;
-        let command_payload =
-            encrypt_command(oldsync_key, target, close_tabs::COMMAND_NAME, &payload)?;
-        self.invoke_command(
-            close_tabs::COMMAND_NAME,
-            target,
-            &command_payload,
-            Some(close_tabs::COMMAND_TTL),
-        )?;
-        self.telemetry.record_command_sent(sent_telemetry);
-        Ok(())
+
+        let sent_telemetry = telemetry::SentCommand::for_close_tabs();
+        let mut urls_to_retry = Vec::new();
+
+        // Sort the URLs shortest to longest, so that we can at least make
+        // some forward progress, even if there's an oversized URL at the
+        // end that won't fit into a single command.
+        let mut urls: Vec<_> = urls.into_iter().map(Into::into).collect();
+        urls.sort_unstable_by_key(String::len);
+
+        while !urls.is_empty() {
+            // If we were asked to close more URLs than will fit in a
+            // single command, chunk the URLs into multiple commands,
+            // packing as many as we can into each. Do this until we've either
+            // drained and packed all the URLs, or we see an oversized URL
+            // that won't fit into a single command.
+            let chunk = match payload_support::try_fit_items(&urls, COMMAND_MAX_PAYLOAD_SIZE.get())
+            {
+                Fit::All => mem::take(&mut urls),
+                Fit::Some(count) => urls.drain(..count.get()).collect(),
+                Fit::None | Fit::Err(_) => {
+                    urls_to_retry.append(&mut urls);
+                    break;
+                }
+            };
+
+            let sent_telemetry = sent_telemetry.clone_with_new_stream_id();
+            let payload = CloseTabsPayload::with_telemetry(&sent_telemetry, chunk);
+
+            let oldsync_key = self.get_scoped_key(scopes::OLD_SYNC)?;
+            let command_payload =
+                encrypt_command(oldsync_key, target, close_tabs::COMMAND_NAME, &payload)?;
+            let result = self.invoke_command(
```

Comment:
(eg, here could be a transient network error, whole the `Fit::None` case is a permanent error, right?)

Reply:
We discussed the reasoning behind this, so can you please add a comment summarizing that here, at least enough so it's clear the behaviour is intentional.

Reply:
> Isn't that impossible in our model? The command+url must be unique.

I think that's a bug in our current model, though?

IIRC, synced tabs don't filter out duplicates, so I thought we agreed on the heuristic that ["if we see N duplicates of a URL in the command, we'll close N instances of that tab"](https://searchfox.org/mozilla-central/rev/87630817796aae3cbceb369a1b642da3e4741745/mobile/android/android-components/components/feature/accounts-push/src/main/java/mozilla/components/feature/accounts/push/CloseTabsCommandReceiver.kt#60-64). There's no way to know _which_ N it is, though.

Did we decide to change those semantics? (It's OK if we did!) I was assuming we'd come back to that, but if we actually think that the "get N-close N" semantics aren't helpful, I can yank that out of Android Components.

Reply:
Done and done!